src/process.cr

```diff
@@ -513,6 +513,156 @@ class Process
     io.close if io
   end
 
+  # Returns the **real** user id of the current process.
+  def self.user_id
+    LibC.getuid.to_i
+  end
+
+  # Returns the **effective** user id of the current process.
+  def self.effective_user_id
+    LibC.geteuid.to_i
+  end
+
+  # Returns the **real** group id of the current process.
+  def self.group_id
+    LibC.getgid.to_i
+  end
+
+  # Returns the **effective** group id of the current process.
+  def self.effective_group_id
+    LibC.getegid.to_i
+  end
+
+  private UID_DEFAULT = if LibC::UidT.new(0).is_a?(Int::Signed)
+                          LibC::UidT.new(-1)
+                        else
+                          LibC::UidT::MAX
+                        end
+  private GID_DEFAULT = if LibC::GidT.new(0).is_a?(Int::Signed)
+                          LibC::GidT.new(-1)
+                        else
+                          LibC::GidT::MAX
+                        end
+
+  # Permanently transition to another account.
+  #
+  # user_ids, group_ids and groups are changed to the account provided.
+  #
+  # Call `chroot` or other privileged operations before calling this method.
+  #
+  # Example:
+  #
+  # ```
+  # user = System::User.find_by name: "crystal"
+  # Process.become user
+  #
+  # user = System::User.find_by name: "crystal"
+  # group = System::Group.find_by name: "wheel"
+  # Process.become user, group # Use a different group other than the user's.
+  # ```
+  def self.become(user : System::User, group : System::Group? = nil)
+    group ||= user.group
+    # TODO: Call setgroups() when available using user.group_ids when available.
+    become_group group.id.to_i
+    become_user user.id.to_i
+  end
+
+  # Changes the real, effective, and saved user ids of the current process.
+  #
+  # Example:
+  #
+  # ```
+  # Process.become_user 0 # Changes real, effective, saved user id to 0 (root).
+  # ```
+  def self.become_user(uid : Int)
+    become_user real: uid, effective: uid, saved: uid
+  end
+
+  # Attempts to change real, effective, and/or saved user id's of the current process.
+  # Uid's not supplied in arguments are kept at their current values (if supported).
+  # Explicit setting of saved id's is not supported on all platforms.
+  #
+  # * *real*: real user id (ruid).
+  # * *effective*: effective user id (euid).
+  # * *saved*: saved user id (suid).
+  #
+  # Example:
+  #
+  # ```
+  # Process.become_user real: 1004, effective: 0 # Changes real and saved user id to 1004 and effective to 0 (root)
+  # Process.become_user effective: 0             # Only changes the effective.
+  # Process.become_user saved: 0                 # Platform specific and may not function.
+  # ```
+  def self.become_user(*, real : Int? = nil, effective : Int? = nil, saved : Int? = nil)
```

Comment:
```suggestion
  def self.become_user(*, real : Int? = nil, effective : Int? = nil, saved : Int = real)
```
`saved` cannot stay unchanged, so it should not accept `nil` which would have different semantics than for the other arguments. Assigning `real` as default value seems logical because that's the default.

Reply:
What happens when setting `effective` without changing `real`?  Requiring `saved` would break that. 

It's a common case. There's a dedicated syscall for it (`seteuid`).  Some services operate in the following manner:
```
while request = ...
  seteuid request.user.id
  open file as user
ensure
  seteuid 0
end
```


Most programs will do one of several things:
* Set them all once when switching away from root.  This is probably the most common case and why Process.become(user) exists.
* Swap euid/ruid back and forth before lowering privileges (setuid or root starting applications).  Example `sudo`.
* Set euid to many users possibly swapping euid/ruid (root running service or setuid).
* I probably missed a few.

Setting `saved` by itself is rare but I wanted to allow it on platforms that support it for the cases when setting `ruid => other, euid => other, suid => current` and switching back later.

Leaving `saved` as the only way back can add additional security since I think it's dropped on exec under some circumstances on some platforms.   Or maybe all platforms.  My memory is foggy on this point as I haven't tried it in > 20 years and don't care to look it up.

Reply:
Oh yeah, I overlooked that case. But `saved = real || effective` should work, right?

Reply:
I just don't want its default value to be `nil` and `nil` having a different meaning than for effective and real.

Reply:
I think I based this on the behavior of using `setreuid`.  Providing `ruid` or `euid` leaves `saved` as whatever `ruid` is, which sometimes means not changing it.  This is probably the most portable option as some platforms don't support `setresuid` and the most secure.

Reply:
Maybe i should document that it attempts to mimic the behavior of setreuid if `saved` is not provided.

Reply:
I'm still not fond of the fact that `nil` argument has a different meaning for real/effective vs. saved

Reply:
If I changed this
`def self.become_user(*, real : Int? = nil, effective : Int? = nil, saved : Int? = nil)`
to this:
`def self.become_user(*, real : Int? = nil, effective : Int? = nil, saved : Int? = real)`

Would that be acceptable?

Reply:
This still means `nil` has the same meaning when passed as argument value like `become_user(real: 1, effective: nil, saved: nil)`.
/cc @ysbaddaden @asterite @RX14 Any input on this?

Reply:
What about two methods?  One that has `setreuid` semantics and another with `setresuid`?
```
def self.become_user(*, real : Int? = nil, effective : Int? = nil)
def self.become_user(*, real : Int? = nil, effective : Int? = nil, saved : Int = nil)
```

Reply:
Yeah, that could work. But it would be `saved : Int = real`, right?

Reply:
Is `become_user` acceptable now?